core/src/main/java/org/mule/runtime/core/internal/processor/chain/ImmutableProcessorChainExecutor.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.processor.chain;
+
+import static java.util.Optional.ofNullable;
+import static org.mule.runtime.core.api.processor.MessageProcessors.processWithChildContext;
+import static reactor.core.publisher.Mono.from;
+import org.mule.runtime.api.lifecycle.Initialisable;
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.api.message.Message;
+import org.mule.runtime.api.metadata.TypedValue;
+import org.mule.runtime.core.api.InternalEvent;
+import org.mule.runtime.core.api.exception.MessagingException;
+import org.mule.runtime.core.api.lifecycle.LifecycleUtils;
+import org.mule.runtime.core.api.processor.MessageProcessorChain;
+import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.internal.processor.EventBasedResult;
+import org.mule.runtime.extension.api.runtime.operation.Result;
+import org.mule.runtime.extension.api.runtime.process.Chain;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * An implementation of {@link Chain} that wraps a {@link Processor} and allows to execute it
+ *
+ * @since 4.0
+ */
+public class ImmutableProcessorChainExecutor implements Chain, Initialisable {
```

Comment:
why is this here instead of extension-support? If the answer is "to keep core for exporting things that shouldn't" is a valid answer. Just making sure the answer is in fact valid

Reply:
uses `InternalEvent` and `MessageProcessorChain`, but they are both exported already. Can be moved

Reply:
if this can be internal then move it to internal. The future of InternalEvent and MessageProcessorChain is uncertain